### DISCUSSION TITLE
how self.camera.pixel_array gets updated with the next frame during the rendering process.

### DISCUSSION THREAD

**Discussion by tianmaoyu** [ManimCommunity/manim#3631] [Q&A] [answered]
###  (Regarding a source code question):  I would like to know how Cairo exports images within Manim.

I know that the image is initially generated based on self.camera.pixel_array, but I am unsure where self.camera.pixel_array gets updated after rendering with Cairo.

When does the **self.camera.pixel_array** get updated with the rendered image when using Cairo for rendering? Below are some locations where I've found the setting of **pixel_array**, however, I'm still unable to find how **self.camera.pixel_array** gets updated with the next frame during the rendering process.

![image](https://github.com/ManimCommunity/manim/assets/10215431/a9547a4d-4287-4f62-982f-9aadb10ea882)

![image](https://github.com/ManimCommunity/manim/assets/10215431/452ad399-1cd6-45e0-b9ca-9e161130e759)

**Comment by behackl** [comment, marked as answer]
See https://github.com/ManimCommunity/manim/blob/77d42d2a46b847b6d5731b4ce254972149c1a426/manim/camera/camera.py#L633 and the lines afterwards.

**Reply by tianmaoyu** [reply]
## When performing various Cairo drawing operations, the data in the **pixel_array** will also be updated, as they share this data.

```python
  surface = cairo.ImageSurface.create_for_data(
            pixel_array,
            cairo.FORMAT_ARGB32,
            pw,
            ph,
        )
```

**Comment by tianmaoyu** [comment]
When performing various Cairo drawing operations, the data in the pixel_array will also be updated, as they share this data.
  surface = cairo.ImageSurface.create_for_data(
            pixel_array,
            cairo.FORMAT_ARGB32,
            pw,
            ph,
        )
